rasterio/_warp.pyx

```diff
@@ -271,6 +272,26 @@ def _reproject(
     cdef GDALTransformerFunc pfnTransformer = NULL
     cdef GDALWarpOptions *psWOptions = NULL
 
+    # If working with identity transform, assume it is crs-less data
+    # and that translating the matrix very slightly will avoid #674
+    eps = 1e-100
+    if src_transform:
+        src_transform = guard_transform(src_transform)
+        # if src_transform is like `identity` with positive or negative `e`,
+        # translate matrix very slightly to avoid #674 and #1272.
+        if src_transform.almost_equals(identity) or src_transform.almost_equals(Affine(1, 0, 0, 0, -1, 0)):
+            src_transform = src_transform.translation(eps, eps)
+    if dst_transform:
+        dst_transform = guard_transform(dst_transform)
+        if dst_transform.almost_equals(identity) or dst_transform.almost_equals(Affine(1, 0, 0, 0, -1, 0)):
+            dst_transform = dst_transform.translation(eps, eps)
+
+    # Convert transforms to GDAL format.
+    if src_transform:
```

Comment:
This block looks partially redundant with the above, it seems like you could merge them?

`guard_transform(src_transform)` and `guard_transform(dst_transform)` are already handled above, as are the `if`s